File: scrooge-generator/src/main/resources/scalagen/finagleClientFunction.scala
private[this] object {{__stats_name}} {
  val RequestsCounter = scopedStats.scope("{{clientFuncNameForWire}}").counter("requests")
  val SuccessCounter = scopedStats.scope("{{clientFuncNameForWire}}").counter("success")
  val FailuresCounter = scopedStats.scope("{{clientFuncNameForWire}}").counter("failures")
  val FailuresScope = scopedStats.scope("{{clientFuncNameForWire}}").scope("failures")
}
{{#functionInfo}}
val {{clientFuncNameForWire}}{{ServiceName}}ReplyDeserializer: Array[Byte] => _root_.com.twitter.util.Try[{{typeName}}] = {
  response: Array[Byte] => {
    val result = decodeResponse(response, {{funcObjectName}}.Result)

    result.firstException() match {
      case Some(exception) => _root_.com.twitter.util.Throw(setServiceName(exception))
      case _ => result.successField match {
        case Some(success) => _root_.com.twitter.util.Return(success)
        case _ => _root_.com.twitter.util.Throw(missingResult("{{clientFuncNameForWire}}"))
      }
    }
  }
}
{{>function}} = {
  {{__stats_name}}.RequestsCounter.incr()
  val inputArgs = {{funcObjectName}}.Args({{argNames}})

  val serdeCtx = new _root_.com.twitter.finagle.thrift.DeserializeCtx[{{typeName}}](inputArgs, {{clientFuncNameForWire}}{{ServiceName}}ReplyDeserializer)
  _root_.com.twitter.finagle.context.Contexts.local.let(
    _root_.com.twitter.finagle.thrift.DeserializeCtx.Key,
    serdeCtx,
    _root_.com.twitter.finagle.thrift.Headers.Request.Key,
    _root_.com.twitter.finagle.thrift.Headers.Request.empty
  ) {
    val serialized = encodeRequest("{{clientFuncNameForWire}}", inputArgs)
    this.service(serialized).flatMap { response =>
      Future.const(serdeCtx.deserialize(response))
    }.respond { response =>
      val responseClass = responseClassifier.applyOrElse(
        ctfs.ReqRep(inputArgs, response),
        ctfs.ResponseClassifier.Default)
      responseClass match {
        case ctfs.ResponseClass.Successful(_) =>
          {{__stats_name}}.SuccessCounter.incr()
        case ctfs.ResponseClass.Failed(_) =>
          {{__stats_name}}.FailuresCounter.incr()
          response match {
            case Throw(ex) =>
              setServiceName(ex)
              {{__stats_name}}.FailuresScope.counter(Throwables.mkString(ex): _*).incr()
            case _ =>
          }
      }
    }
  }
}
{{/functionInfo}}

File: scrooge-generator/src/main/resources/javagen/finagleClientFunction.java

private __Stats _{{__stats_name}};

private __Stats {{__stats_name}}() {
  if (_{{__stats_name}} == null) {
    _{{__stats_name}} = new __Stats("{{clientFuncNameForWire}}");
  }
  return _{{__stats_name}};
}

{{#headerInfo}}{{>header}}{{/headerInfo}} {
  {{__stats_name}}().requestsCounter.incr();

  Future<{{type}}> rv = this.service.apply(encodeRequest("{{clientFuncNameForWire}}", new {{ArgsStruct}}({{argNames}}))).flatMap(new Function<byte[], Future<{{type}}>>() {
    public Future<{{type}}> apply(byte[] in) {
      try {
        {{ResultStruct}} result = decodeResponse(in, {{ResultStruct}}.CODEC);

{{#hasThrows}}
        Exception exception = null;
{{#throws}}
        if (exception == null && result.{{throwName}}.isDefined()) exception = result.{{throwName}}.get();
{{/throws}}
        if (exception != null) return Future.exception(exception);
{{/hasThrows}}

{{#isVoid}}
        return Future.value(null);
{{/isVoid}}
{{^isVoid}}
        if (result.success.isDefined()) return Future.value(result.success.get());
        return Future.exception(missingResult("{{clientFuncNameForWire}}"));
{{/isVoid}}
      } catch (TException e) {
        return Future.exception(e);
      }
    }
  }).rescue(new Function<Throwable, Future<{{type}}>>() {
    public Future<{{type}}> apply(Throwable t) {
      if (t instanceof SourcedException) {
        ((SourcedException) t).serviceName_$eq(FinagledClient.this.serviceName);
      }
      return Future.exception(t);
    }
  });

  rv.addEventListener(new FutureEventListener<{{type}}>() {
    public void onSuccess({{type}} result) {
      {{__stats_name}}().successCounter.incr();
    }

    public void onFailure(Throwable t) {
      {{__stats_name}}().failuresCounter.incr();
      {{__stats_name}}().failuresScope.counter0(t.getClass().getName()).incr();
    }
  });

  return rv;
}
